osf/external/gravy_valet/auth_helpers.py

```diff
@@ -0,0 +1,132 @@
+import base64
+import email.utils
+import hashlib
+import hmac
+import re
+import typing
+import urllib
+
+from django.utils import timezone
+
+from osf.models import OSFUser, AbstractNode
+from osf.utils import permissions as osf_permissions
+from website import settings
+
+_AUTH_HEADER_REGEX = re.compile(
+    r'^HMAC-SHA256 SignedHeaders=(?P<headers>[\w;-]*)&Signature=(?P<signature>[^\W_]*$)'
+)
+
+
+def _sign_message(message: str, hmac_key: str = None) -> str:
+    key = hmac_key or settings.DEFAULT_HMAC_SECRET
+    encoded_message = base64.b64encode(message.encode())
+    return hmac.new(
+        key=key.encode(), digestmod=hashlib.sha256, msg=encoded_message
+    ).hexdigest()
+
+
+def _get_signed_components(
+    request_url: str, request_method: str, body: typing.Union[str, bytes], **additional_headers
+) -> typing.Tuple[typing.List[str], typing.Dict[str, str]]:
+    parsed_url = urllib.parse.urlparse(request_url)
+    if isinstance(body, str):
+        body = body.encode()
+    content_hash = hashlib.sha256(body).hexdigest() if body else None
+    auth_timestamp = email.utils.format_datetime(timezone.now())
```

Comment:
hmm, since this isn't actually going in the `Date` header, could we use a sane date serialization instead?

Reply:
That would be another thing to sync across GV and OSF, whereas this is at least a standard format (even if it's dumb).

Reply:
isoformat! (can't (easily) parse osf-side, but don't need to for tests -- we only worry about validating the signature)